Option to save single shots for coherence signal routines

stavros11 opened the pull request:
Adds the option to save single shots for coherence routines in case we need it for monitoring. We should also:

- [x] propage to spin echo
- [ ] check if we can get errors for the fitted T1, T2 values even when using the `_signal` routines (in case classification is not very good and we cannot use probabilities)

andrea-pasquale reviewed (commented):
Thanks @stavros11, I still need to test on hardware but it looks ok.
I have just a few minor comments.
It could be nice to find a way to reduce code duplication (cc @hay-k), feel free to open an issue about it and we can address it later on.

stavros11 commented on tests/runcards/protocols.yml:
```diff
@@ -304,6 +304,16 @@ actions:
       delay_before_readout_step: 2000
       nshots: 1024
 
+  - id: t1_signal
```
Thanks, fixed in d522466971b8a6c2748f27657f714c3fcc1e7b7b.

andrea-pasquale reviewed (approved):
Thanks @stavros11.

stavros11 commented:
@andrea-pasquale thank you for reviewing. Sorry, I had to re-request review because in f63391a02f6968220934e59b86fa62c415f96e48 I dropped the new classes I added (`T1SignalSingleShotData`, etc.) because I realized that fitting was not working as expected when using `qq fit` on data that already exist (I have only tested with `qq auto` before). I also added the `pcov` error returned by `curve_fit` in the fitting results, so that we have some error estimate for the signal routines too.

andrea-pasquale commented:
Thanks @stavros11, don't worry about asking again for a review :)
I saw that tests were failing because something was wrong in the table generation for the HTML report.
I pushed something to fix it, quickly. Let me know if this is what you had in mind.
Considering that now we are also showing the error we could also add the $\chi^2$ just to have a metric for the fit.

stavros11 commented:
Thanks @andrea-pasquale. I knew tests were failing and I could not fix them in the first quick look, so I was planning to come back at it later. Thanks for fixing it.

I already used this branch for the monitoring of the single qubit and it worked, but since now some things changed I will try to test on hardware again before merging (once we reach these routines for the 5q chip).

stavros11 commented:
> Considering that now we are also showing the error we could also add the χ2 just to have a metric for the fit.

That would also be great, however I am not sure how I would do it. For the t1, t2 routines (with discrimination) you are using an error calculated as:
https://github.com/qiboteam/qibocal/blob/9d0042da1d67c348ec3f4854c4a53614c74d666b/src/qibocal/protocols/characterization/coherence/t2.py#L100
in the calculation of chi2. I am not sure what is the corresponding formula (if there is one) for the signal routines where we are using voltages.

andrea-pasquale commented:
> > Considering that now we are also showing the error we could also add the χ2 just to have a metric for the fit.
> 
> That would also be great, however I am not sure how I would do it. For the t1, t2 routines (with discrimination) you are using an error calculated as:
> 
> https://github.com/qiboteam/qibocal/blob/9d0042da1d67c348ec3f4854c4a53614c74d666b/src/qibocal/protocols/characterization/coherence/t2.py#L100
> 
> in the calculation of chi2. I am not sure what is the corresponding formula (if there is one) for the signal routines where we are using voltages.

Yeah, this can be tricky. Either we assume that the distribution is gaussian and we use the standard deviation as error or I don't know... 

We can postpone it for a later PR.

Edoardo-Pedicillo commented:
> > Considering that now we are also showing the error we could also add the χ2 just to have a metric for the fit.
> 
> That would also be great, however I am not sure how I would do it. For the t1, t2 routines (with discrimination) you are using an error calculated as:
> 
> https://github.com/qiboteam/qibocal/blob/9d0042da1d67c348ec3f4854c4a53614c74d666b/src/qibocal/protocols/characterization/coherence/t2.py#L100
> 
> 
> in the calculation of chi2. I am not sure what is the corresponding formula (if there is one) for the signal routines where we are using voltages.

The evaluation of the errors is not trivial in the signal case, we tried different approaches but we always get very small error bars and this means really high chi squares.  Instead of showing the standard deviation in the plots, that in this case the usage is debatable, we could just show the variability of the data with a boxplot in the report.

About the chi square @igres26 suggested to evaluate just the distance between the predictions and the data instead.
I think in this way we could avoid the error evaluation.

andrea-pasquale commented:
@stavros11 if on hardware there are no issues feel free to merge

stavros11 commented:
> @stavros11 if on hardware there are no issues feel free to merge

Thanks @andrea-pasquale. It seems to work on hardware:

* T1: http://login.qrccluster.com:9000/jFACcs_XScaYRx_pNyyjuQ==
* T2: http://login.qrccluster.com:9000/iAGa332uTZad0m2AOyNcIQ==

T2 doesn't look good but I don't think it is related to this PR. CI is also passing, so I will merge.
